Learning function for biased noise

Misty-W opened the pull request:
Description
-----------

Add high-level learning function for learning-based PEC. 

API-doc is updated, user guide update to follow in separate issue/PR : #1365

Fixes #821

Checklist
---------

Check off the following once complete (or if not applicable) after opening the PR. The PR will be reviewed once this checklist is complete and all tests are passing.

- [x] I added unit tests for new code.
- [x] I used [type hints](https://www.python.org/dev/peps/pep-0484/) in function signatures.
- [x] I used [Google-style](https://google.github.io/styleguide/pyguide.html#383-functions-and-methods) docstrings for functions.
- [x] I [updated the documentation](../blob/master/docs/CONTRIBUTING_DOCS.md) where relevant.
- [x] Added myself / the copyright holder to the AUTHORS file

If some items remain, you can mark this a [draft pull request](https://github.blog/2019-02-14-introducing-draft-pull-requests/).

License
-------

- [x] I license this contribution under the terms of the GNU GPL, version 3 and grant Unitary Fund the right to provide additional permissions as described in section 7 of the GNU GPL, version 3.


Tips
----

- If the validation check fails:

    1. Run `make check-types` (from the root directory of the repository) and fix any [mypy](https://mypy.readthedocs.io/en/stable/) errors.

    2. Run `make check-style` and fix any [flake8](http://flake8.pycqa.org) errors.

    3. Run `make format` to format your code with the [black](https://black.readthedocs.io/en/stable/index.html) autoformatter.

  For more information, check the [Mitiq style guidelines](https://mitiq.readthedocs.io/en/stable/contributing.html#style-guidelines).

- Write "Fixes #XYZ" in the description if this PR fixes Issue #XYZ.

Misty-W commented:
As demonstrated in PR #1356, the learning and loss functions can be made public because multiple frontends are supported.

andreamari reviewed (commented):
Hi @Misty-W, I left some comments and suggestions.
Feel free to apply them in ways that are different from the suggested changes. 
In particular, I am not sure if the suggested docstrings are correct.

My two main concerns are:

- We need to be confident that optimization works as expected. (See comments on test).
- We are not testing different frontends. Is it possible to have a lightweight test for at least a non-Cirq frontend? In this case, testing that it just runs is fine. No need to test the correct learning of the noise parameters, I guess.

Misty-W commented on mitiq/pec/representations/tests/test_learning.py:
```diff
@@ -135,3 +138,38 @@ def noisy_execute(circ: Circuit) -> np.ndarray:
+        eta0=(1 + offset) * eta,
+        observable=observable,
+    )
+    assert np.isclose(epsilon_opt, epsilon, rtol=1e-01, atol=1e-02)
```
I looked into this some more, looks like the optimization is running but not converging to the minimum.

I'm trying out some different settings (`number_of_training_circuits`,`fraction_non_clifford`, `num_samples`, `offset` (for initial parameters), etc. to see if I can find a better combination.

Misty-W commented on mitiq/pec/representations/learning.py:
```diff
@@ -17,14 +17,86 @@
+        [ideal_executor.evaluate(t, observable) for t in training_circuits]
+    )
+    # Set number of samples used to calculate mitigated value in loss function
+    pec_kwargs = {"num_samples": 50}
```
Yes, that's fixed now.

Misty-W commented on mitiq/pec/representations/learning.py:
```diff
@@ -17,14 +17,86 @@
+    training_circuits = generate_training_circuits(
+        circuit=circuit,
+        num_training_circuits=num_training_circuits,
+        fraction_non_clifford=0,
```
Makes sense, I added `fraction_non_clifford` as an argument of `learn_biased_noise_parameters`. 

I'm also trying non-zero values of `fraction_non_clifford` in the tests to see if that helps the optimization complete successfully.

Misty-W commented on mitiq/pec/representations/learning.py:
```diff
@@ -17,14 +17,86 @@
+            observable,
+        ),
+        method="BFGS",
+    )
```
I implemented the option of `minimize_kwargs`, as I think flexibility for the user is key, especially at this early stage.

nathanshammah commented:
- 1. Seed in tests may help with avoid stochastic fails in continuous integration and pin down the problem
- 2. Many files are changed. Rebasing against master may help solve this?
- 3. Simplifying the problem used in test (e.g., simpler quantum circuit) may help prevent having the optimization fail/be unstable?

Misty-W commented:
> * 1. Seed in tests may help with avoid stochastic fails in continuous integration and pin down the problem
> * 2. Many files are changed. Rebasing against master may help solve this?
> * 3. Simplifying the problem used in test (e.g., simpler quantum circuit) may help prevent having the optimization fail/be unstable?

* 1. Implemented
* 2. Fixed with rebase
* 3. Pinpointed problem to nonsensically high PEC - mitigated expectation values ~30,000 for a couple of the training circuits. If it's not clear what's causing this on the existing circuit I'll try a simpler one.

nathanshammah commented:
@Misty-W great on the update. +1 on trying a simpler circuit.

Misty-W commented:
@andreamari @nathanshammah, some good news: tests for Rx, Rz, and CNOT, for `epsilon` 0.05 and 0.1 are passing!
The bad news is the execution time is still too long, especially in CI. I'm not sure why the Linux tests in CI were about 1h longer than my local run, and the non-Linux tests were even slower. I also had to re-run the non-Linux tests, but I believe that was due to the long execution time.

To do: 
1. Combine test functions for CNOT and rotation gates- right now they are separate because one executor gives the correct result for CNOT and the other gives the correct result for rotation gates. 
2. Find ways of shortening execution time --> suggestions welcome! 
    Here are the ideas I have so far:
    a.  Making the executor used in the Rx/Rz test work correctly for 2-qubit gates as well, because it doesn't iterate over the circuit each time to check whether/where to insert the noisy operation
    b.   Try different insertion strategies for the noisy operation to see if they make a difference in execution time- I don't expect them to make a difference in this case, but it might be worth a try.
    c.  The way I wrote the Rx and Rz tests makes them redundant because I'm just inserting one custom noisy operation defined in terms of the biased noise channel. Unless I'm forgetting something, I think we can eliminate Rx or Rz without losing actual coverage.
3. Other cleanup: change hardcoded random seed for `generate_training_circuits` into an argument in `learn_biased_noise_parameters` and clean up `gitignore`

andreamari commented:
@Misty-W 
> 2. Find ways of shortening execution time --> suggestions welcome!

Have you tried setting initial conditions for epsilon closer or (even equal) to the theoretical optimal values. If execution time is so large, we may need similar _extreme_ workarounds.

Misty-W commented:
> @Misty-W
> 
> > 2. Find ways of shortening execution time --> suggestions welcome!
> 
> Have you tried setting initial conditions for epsilon closer or (even equal) to the theoretical optimal values. If execution time is so large, we may need similar _extreme_ workarounds.

@andreamari, indeed the intention of the test is to only iterate a few times by setting the initial value of the noise strength close to the true value. An initial offset of 1% from the true value works for CNOT but in the case of Rx and Rz causes ` assert abs(epsilon_opt - epsilon) < abs(epsilon0 - epsilon)` to fail. I'm experimenting with using different offset values for different gates. 

We could also consider an alternate condition that allows for smaller offsets even if the optimization is not that accurate, but we run the risk of missing situations where the optimization doesn't run but the test still passes, as happened when I first opened this PR.

Misty-W commented:
@andreamari, tests are passing and execution time is looking much better! 

Using 
- `Nelder-Mead` solver instead of `BFGS`
- `num_samples = 50` instead of `num_samples = 5000` 
- Applied initial noise strength offset in the opposite direction, i.e. `epsilon0 = (1 - offset) * epsilon` instead of `epsilon0 = (1 + offset) * epsilon`

I also removed the Qiskit test cases for now as they are still slow (probably due to a "mismatched" application of noise in the executor). Unless we find a blocking issue regarding the `test_learn_biased_noise_parameters` cases, I'll work on reducing the Qiskit test execution time.

andreamari reviewed (commented):
Thanks @Misty-W, there is an important comment about the test condition which may change the results, plus other minor things. 

Nice to see a good execution time!  Feel free to even drop rotations to further reduce the execution time, if necessary.

Misty-W commented on mitiq/pec/representations/learning.py:
```diff
@@ -15,16 +15,107 @@
 
 
-def _biased_noise_loss_function(
+def learn_biased_noise_parameters(
```
Yes, updated function name to `learn_depolarizing_noise_parameter`

Misty-W commented on mitiq/pec/representations/learning.py:
```diff
@@ -81,6 +172,6 @@ def _biased_noise_loss_function(
-    )
+    return np.sum(
+        abs(mitigated_values.reshape(-1, 1) - ideal_values.reshape(-1, 1)) ** 2
+    ) / len(training_circuits)
```
Good catch, this works!

Misty-W commented on mitiq/pec/representations/tests/test_learning.py:
```diff
@@ -135,3 +142,46 @@ def noisy_execute(circ: Circuit) -> np.ndarray:
+        observable=observable,
+    )
+    assert success
+    assert abs(epsilon_opt - epsilon) < offset
```
I knew it had to be too good to be true! But... even after updating the test condition to `abs(epsilon_opt - epsilon) < offset * epsilon`, with a modest increase in samples and offset, the tests are passing yet again!

I took out the rotations to save execution time, especially in light of the increased samples for CNOT.

Misty-W commented on mitiq/pec/representations/tests/test_learning.py:
```diff
@@ -135,3 +142,46 @@ def noisy_execute(circ: Circuit) -> np.ndarray:
+    eta = 0
+    index = operations[0]
+    op = operations[1]
+
+    def noisy_execute(circ: Circuit) -> np.ndarray:
```
Good to have a note about this here, in case there's a need to generalize this test in the future.

andreamari commented on mitiq/pec/representations/learning.py:
```diff
@@ -15,16 +15,109 @@
+    observable: Optional[Observable] = None,
+    **minimize_kwargs: Dict["str", Any],
+) -> Tuple[bool, float]:
+    r"""This function learns the depolarizing noise parameter (epsilon)
+    associated to a set of input operations. The learning process is based on
+    the execution of a set of training circuits on a noisy backend and on a
+    classical simulator. The training circuits are near-Clifford approximations
+    of the input circuit. A depolarizing noise model characterization is
+    assumed.
```
What about if we mention somewhere in the docstring something like this?

_Please note that using this function is quite difficult. One of the main difficulties is setting a good value of `num_samples` in the PEC options (`pec_kwargs`).  Setting a small value of `num_samples` is typically necessary to obtain a reasonable execution time. On the other hand, with a small number of PEC samples, the statistical error can be so large that the optimization process fails._

There is probably a better way to rephrase this. But it can be worth giving a similar warning to user before they go through the same problems that we had in our CI tests.

Misty-W replied:
Good point, I added the following paragraph just after the text you marked?

> Note: using this function may require some tuning. One of the main
> challenges is setting a good value of `num_samples` in the PEC options
> `pec_kwargs`. Setting a small value of `num_samples` is typically
>  necessary to obtain a reasonable execution time. On the other hand,
>  using a number of PEC samples that is too small can result in a large
>  statistical error, ultimately causing the optimization process to fail.

andreamari reviewed (commented):
Thanks @Misty-W!
Adding non-cirq tests would make CI too slow? 
What about if we add a non-cirq test just for `biased_noise_loss_function`? Still too slow?

A few other comments are below.

Misty-W commented:
> Thanks @Misty-W! Adding non-cirq tests would make CI too slow? What about if we add a non-cirq test just for `biased_noise_loss_function`? Still too slow?
> 
> A few other comments are below.

Thanks for the suggestion @andreamari, non-Cirq test on `biased_noise_loss_function` is a good compromise.

Misty-W commented on mitiq/pec/representations/learning.py:
```diff
@@ -15,16 +15,109 @@
+            observable,
+        ),
+        method="Nelder-Mead",
+        **minimize_kwargs,
```
I added an argument `solver` to `learn_depolarizing_noise_parameter` with `"Nelder-Mead"` as default, since Nelder-Mead works reliably for the tests, unlike BFGS and others.

Misty-W commented:
@andreamari, thanks for the review yesterday! Anything else to address before merging?

andreamari commented:
> @andreamari, thanks for the review yesterday! Anything else to address before merging?

After small changes discussed during the Mitiq call, this is ready to go for me!

andreamari commented on mitiq/pec/representations/learning.py:
```diff
@@ -155,13 +157,14 @@ def biased_noise_loss_function(
-            the `QuantumResult` returned by `executor` is used to compute the
-            expectation of the observable.
+            If None, the ``executor`` must return an expectation value.
+            Otherwise the ``QuantumResult`` returned by `'executor`` is used to
```
```suggestion
            Otherwise the ``QuantumResult`` returned by ``executor`` is used to
```

andreamari reviewed (approved):
LGTM!
Great to finally have learning based PEC in Mitiq! 
Let's wait all test pass before merging.

Misty-W commented:
Merged at last! Many thanks for your guidance and feedback @andreamari and @nathanshamah! 

cc: @willzeng, @natestemen

willzeng commented:
Great work and fantastic to see this feature now integrated in.

On Fri, Sep 23, 2022 at 3:07 PM Misty Wahl ***@***.***> wrote:

> Merged at last! Many thanks for your guidance and feedback @andreamari
> <https://github.com/andreamari> and @nathanshamah!
>
> cc: @willzeng <https://github.com/willzeng>, @natestemen
> <https://github.com/natestemen>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/unitaryfund/mitiq/pull/1358#issuecomment-1256576962>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABHZDAX7ZADFSHHD5P2QRX3V7X5Y5ANCNFSM5YYRZXJQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
https://unitary.fund/
Creating a quantum technology ecosystem that benefits the most people.
